research/job_referencing/index.md

```diff
@@ -0,0 +1,239 @@
+    self._srpm_builded = False
+    self._copr_builded = False
+
+  @reacts_to(event=CoprBuildEndEvent)
+  @workflow.signal
+  def submit_srpm_build_ended():
+    if event.chroot == COPR_SRPM_CHROOT:
+      self._srpm_builded = True
+
+  @reacts_to(event=CoprBuildEndEvent)
+  @workflow.signal
+  def submit_copr_build_ended():
+    if event.chroot != COPR_SRPM_CHROOT:
+      self._copr_builded = True
+
+  @workflow.run
+  async def run(self):
+
+    srpm_activity_handle = workflow.start_activity(
+      submit_build_srpm,
+      start_to_close_timeout=timedelta(seconds=1),
+      ...
+      )
+    await workflow.wait_condition(lambda: self._srpm_builded)
+
+    update_ux_handle = workflow.start_activity(
+      update_ux,
+      start_to_close_timeout=timedelta(seconds=1),
+      ...
+      )
+
+    build_copr_handle = workflow.start_activity(
+      build_copr,
+      start_to_close_timeout=timedelta(seconds=1),
+      ...
+      )
+    await workflow.wait_condition(lambda: self._copr_builded)
+
+    update_ux_handle = workflow.start_activity(
+      update_ux,
+      start_to_close_timeout=timedelta(seconds=1),
+      ...
+      )
+```
+
+As far as I understand from the `temporalio` documentation, the framework is in charge of putting this `CoprBuildHandler workflow` in a sleeping queue while it is waiting for a signal and waking up it later. So we don't need to exit the workflow and the order of the activities is well visible in the `workflow` code.
+
+There is another small improvement I see using a solution like this one instead of doing what we are already doing.
+I don't think that the `CoprBuildEndHandler` should deal with the `TestingFarmHandler` and viceversa, this is dangerous, and the more dependencies we will introduce the more we will need this dangerous cross referencing.
+
+Something above them should be able to orchestrate and know the jobs, and I think a `workflow` is exactly what we are missing here.
+
+## Conclusion
+
+From my point of view we have two possible ways to implement job referencing:
+
+1.  We can keep doing what we are already doing: use Celery and spawn different tasks when we need them.
+
+    PRO: this is the quickest solution
+
+    CON: it does not scale well, every new reference makes our code less readable/testeable and makes managing errors harder
+
+2.  Introduce a workflow engine with support for tasks communication
+
+    PRO: it should make our code more readable/testeable and we should be able to better manage error handling
+
+    CON: it will take us far more time
+
+An hybrid solution could be: go with solution 1 and implement just those job referencing we believe are actually/really needed.
+If we realize we would need other job referencing in future, we want to make job referencing more generic or we want the user to be able to build his own workflow then we need to start studying a workflow manager (like temporal) and planning a change in our core code base to go with solution 2.
```

Comment:
I would probably keep in mind the need to deprecate whatever comes out of the first solution.
Yes, we can adjust the configs almost automatically, but still.